builtin/v13/check.go

```diff
@@ -288,6 +288,61 @@ func CheckDealStatesAgainstSectors(acc *builtin.MessageAccumulator, minerSummari
 		acc.Require(deal.SlashEpoch <= sectorDeal.SectorExpiration,
 			"deal state slashed at %d after sector expiration %d for miner %v",
 			deal.SlashEpoch, sectorDeal.SectorExpiration, deal.Provider)
+
+		acc.Require((deal.SectorNumber == sectorDeal.SectorNumber) || (deal.SectorNumber == 0 && deal.LastUpdatedEpoch != -1),
+			"deal sector number %d does not match sector %d for miner %v (ds: %#v; ss %#v)",
+			deal.SectorNumber, sectorDeal.SectorNumber, deal.Provider, deal, sectorDeal)
+	}
+
+	// HAMT[ActorID]HAMT[SectorNumber]SectorDealIDs
+	marketDealToSector := make(map[abi.DealID]abi.SectorID)
+
+	for sectorID, dealIDs := range marketSummary.ProviderSectors {
+		maddr, err := address.NewIDAddress(uint64(sectorID.Miner))
+		if err != nil {
+			acc.Addf("error creating ID address: %v", err)
+			continue
+		}
+
+		minerSummary, found := minerSummaries[maddr]
+		if !found {
+			acc.Addf("provider %v for sector %v not found among miners", sectorID, sectorID)
+			continue
+		}
+
+		sector := minerSummary.SectorsWithDeals[sectorID.Number]
+		acc.Require(sector, "sector %v not found in miner %v for deals %v", sectorID, maddr, dealIDs)
+
+		for _, dealID := range dealIDs {
+			_, found := minerSummary.Deals[dealID]
+			acc.Require(found, "deal %d not found in miner %v for sector %v", dealID, maddr, sectorID)
+
+			_, found = marketDealToSector[dealID]
+			acc.Require(!found, "deal %d found in multiple sectors", dealID)
+
+			marketDealToSector[dealID] = sectorID
+		}
+	}
+
+	for sectorID, ids := range marketSummary.ProviderSectors {
+		acc.Require(len(ids) > 0, "no deal ids in sector %v", sectorID)
+
+		for _, dealID := range ids {
+			deal, found := marketSummary.Deals[dealID]
+			acc.Require(found, "providersectors %v sector deal %d not found in market", sectorID, dealID)
+			if !found {
+				continue
+			}
+
+			provID, err := address.IDFromAddress(deal.Provider)
+			if err != nil {
+				acc.Addf("error creating ID address: %v", err)
+				continue
+			}
+
+			acc.Require(abi.ActorID(provID) == sectorID.Miner, "deal %d has provider %v, expected %v", dealID, deal.Provider, sectorID)
+			acc.Require(deal.SectorNumber == sectorID.Number, "deal %d has sector number %d, expected %d", dealID, deal.SectorNumber, sectorID.Number)
+		}
```

Comment:
This check is internal to the market actor now, so should move into market/invariants.go.

Reply:
Moved into market checks